### DISCUSSION TITLE
Parent constructor determination

### DISCUSSION THREAD

**Discussion by yahoo0742** [dotnet/csharplang#6168] [Language Ideas]
Hi, I'm wondering whether my requirement below can be met by the language itself instead of I adding some wrappers.
```
class X {}

class Y {}

class A {
    public A(X x) {}
    public A(Y y) {}
    // don't want to add another constructor with an optional argument
}

class B : A {
    private static y = new Y();
    public B(X x=null) : base(x!=null? (X)x: (Y)y) {} // I want to do this but seems can't now!
}
```
Thanks!

**Comment by TahirAhmadov** [comment]
That's not possible they way you envisioned it now. The challenge is that C# tries to ensure that each instance is initialized correctly, and mandates that constructors are called in a certain order - base has to be called first, etc. This means, it's difficult to provide flexibility to call one of many possible base ctors at runtime (but probably not impossible with a language change). 

For your case, you can do:
```cs
class X {}

class Y {}

class A {
    public A(X x) : this(x, null) {}
    public A(Y y) : this(null, y) {}
    protected A(X? x = null, Y? y = null) { ... }
}

class B : A {
    private static y = new Y();
    public B(X x=null) : base(x, y) {}
}
```

**Reply by yahoo0742** [reply]
@TahirAhmadov Thanks for your reply. You can see in my comments Im reluctant to add another constructor with an optional argument.
I'm wondering whether the language can support it. If can't, why.
I'm afraid the order of executing base constructors or which constructor won't be a challenge, because compilers should be able to know whether there is any invalid statements before calling base constructors. Perhaps the challenge comes from how the the base constructors is referred/pointed in a virtual table or something like that.

**Reply by TahirAhmadov** [reply]
I think it may be possible to change the language to allow a construct like you described when calling base constructors; however, your syntax is probably too difficult to understand, and even if good syntax is devised, this would be a very low priority item - given that adding another constructor is trivial.

**Reply by HaloFour** [reply]
The language and runtime probably could support it, but the question is whether or not it's worth the effort to make that happen.  This would seemingly be an uncommon scenario and as the author of the class it's trivial to solve the problem through other means.

IMO it wouldn't be worth considering this on it's own.  There have been [other discussions](https://github.com/dotnet/csharplang/discussions/4827) about allowing the `base(...)` constructor call to appear within the body of the constructor to better support some scenarios with validation or initialization.  If that were to be considered then it's more a question of where within the constructor body the base constructor calls must appear, and whether you could call different base constructors in different conditional branches, as long as the compiler can guarantee that exactly one base constructor is called.  Something like this:

```cs
class B : A {
    private static y = new Y();
    
    public B(X x=null) {
        if (x != null) {
            base(x);
        } else {
            base(y);
        }
    }
}
```

Although I don't know how fancy the validation in the runtime is, or if that even matters.

**Reply by yahoo0742** [reply]
Thanks @HaloFour I completely support the option you listed above.  I didnt get the point in that thread. Is this option supported or not? why?
